File: JavaScript/877-Stone Game.js
/**
 * @param {number[]} piles
 * @return {boolean}
 */
var stoneGame = function(piles) {
    const N = piles.length;
    let memo = new Map();
    let dfs = (start, end)=>{
        if (start > end) return 0;
        if (memo.has(`${start}-${end}`)){
            return memo.get(`${start}-${end}`);
        }
        memo.set(`${start}-${end}`, Math.max(piles[start] - dfs(start+1, end), piles[end]-dfs(start, end-1)));
        return memo.get(`${start}-${end}`);
    }
    
    return dfs(0,N-1) > 0;
};

File: C++/877-Stone Game.cpp
class Solution {
public:
    bool stoneGame(vector<int>& piles) {
        int N = piles.size();
        vector<vector<int>> dp(N, vector<int>(N, INT_MAX));

        // dp[i][j] : largest score difference

        for (int i=0; i<N; i++) {
            dp[i][i] = piles[i];
        }

        for (int len=1; len<N; len++) {
            for (int i=0; i<N; i++) {
                int j = i+len-1;
                if (i+len-1 < N) break;
                dp[i][j] = max(piles[i] - dp[i-1][j], piles[j] - dp[i][j-1]);
            }
        }

        return dp[0][N-1];
    }
};


class Solution {
public:
    bool stoneGame(vector<int>& piles) {
        // when the length is even, alice can pick all odd-index stones or all even-index stones.
        // Note that the total number of stones is odd, a tie is impossible

        return true;
    }
};